src/libraries/System.Net.Security/src/System/Net/Security/SslStream.IO.cs

```diff
@@ -763,13 +765,16 @@ private SecurityStatusPal DecryptData(int frameSize)
         private async ValueTask<int> ReadAsyncInternal<TIOAdapter>(Memory<byte> buffer, CancellationToken cancellationToken)
             where TIOAdapter : IReadWriteAdapter
         {
-            if (Interlocked.Exchange(ref _nestedRead, 1) == 1)
+            // Throw first if we already have exception.
+            // Check for disposal is not atomic so we will check again bellow.
+            ThrowIfExceptionalOrNotAuthenticated();
+
+            if (Interlocked.CompareExchange(ref _nestedRead, 1, 0) != 0)
```

Comment:
Might be worth adding a comment to the equivalent code in WriteAsync as to why it doesn't get similar treatment